src/state/block_tree.rs

```diff
@@ -0,0 +1,484 @@
+        }
+
+        let mut siblings = self
+            .children(&block.justify.block)
+            .unwrap_or(ChildrenList::default());
+        siblings.push(block.hash);
+        wb.set_children(&block.justify.block, &siblings)?;
+
+        self.write(wb);
+
+        Ok(())
+    }
+
+    pub fn set_highest_qc(&mut self, qc: &QuorumCertificate) -> Result<(), BlockTreeError> {
+        let mut wb = BlockTreeWriteBatch::new();
+        wb.set_highest_qc(qc)?;
+        self.write(wb);
+        Ok(())
+    }
+
+    pub fn set_highest_tc(&mut self, tc: &TimeoutCertificate) -> Result<(), BlockTreeError> {
+        let mut wb = BlockTreeWriteBatch::new();
+        wb.set_highest_tc(tc)?;
+        self.write(wb);
+        Ok(())
+    }
+
+    pub fn set_highest_view_entered(&mut self, view: ViewNumber) -> Result<(), BlockTreeError> {
+        let mut wb = BlockTreeWriteBatch::new();
+        wb.set_highest_view_entered(view)?;
+        self.write(wb);
+        Ok(())
+    }
+
+    pub fn set_highest_view_voted(&mut self, view: ViewNumber) -> Result<(), BlockTreeError> {
+        let mut wb = BlockTreeWriteBatch::new();
+        wb.set_highest_view_voted(view)?;
+        self.write(wb);
+        Ok(())
+    }
+
+    /* ↓↓↓ For deleting abandoned branches in insert_block ↓↓↓ */
+
+    /// Delete the "siblings" of the specified block, along with all of its associated data (e.g., pending
+    /// app state updates). Siblings
+    /// here refer to other blocks that share the same parent as the specified block.
+    ///
+    ///  # Precondition
+    /// Block is in its parents' (or the genesis) children list.
+    ///
+    /// # Error
+    /// Returns an error if the block is not in the block tree, or if the block's parent (or genesis) does not have a
+    /// children list.
+    pub fn delete_siblings(
+        &mut self,
+        wb: &mut BlockTreeWriteBatch<K::WriteBatch>,
+        block: &CryptoHash,
+    ) -> Result<(), BlockTreeError> {
+        let parent_or_genesis = self.block_justify(block)?.block;
+        let parents_or_genesis_children = self.children(&parent_or_genesis)?;
+        let siblings = parents_or_genesis_children
+            .iter()
+            .filter(|sib| *sib != block);
+        for sibling in siblings {
+            self.delete_branch(wb, sibling);
+        }
+
+        wb.set_children(&parent_or_genesis, &ChildrenList::new(vec![*block]))?;
+        Ok(())
+    }
+
+    /// Deletes all data of blocks in a branch starting from (and including) a given root block.
+    fn delete_branch(&mut self, wb: &mut BlockTreeWriteBatch<K::WriteBatch>, root: &CryptoHash) {
+        for block in self.blocks_in_branch(*root) {
+            wb.delete_children(&block);
+            wb.delete_pending_app_state_updates(&block);
+            wb.delete_block_validator_set_updates(&block);
+
+            if let Ok(Some(data_len)) = self.block_data_len(&block) {
```

Comment:
This `if let` statement quietly ignores the case where `block_data_len` returns an `Err`. Very dangerous.

Reply:
Will fix with https://github.com/parallelchain-io/hotstuff_rs/issues/44